### ISSUE TITLE
Feature request: Option to make Object Types managed by the map itself

### ISSUE THREAD

**Issue by SSBMTonberry** [bjorn/tiled#2154]
Hello!
First I would like to thank you for this awesome tool! I've been using it for many years, and it is by far the most amazing program of its kind!

There is however something I've been missing for the longest time.
I've seen a number of issues regarding "Option to make Object Types managed by the map itself" (or something similar) not being a feature. Ref: #1987 #1609 #228 #1263 #1598

I am one of the guys that really would benefit of either having all types in the Object Types Editor handled by the map itself, or at the very least have the possibility to export the actual properties of the Object Types from the Object Types Editor even when the value is the same as the default. I've used Tiled for quite some time, and have been missing such a feature in many years.

I understand that the way it works right now is intentional, but it is not intuitive for me that I have to access Object Types by an external file in my projects, when everything else can be resolved by simply loading the map file. Let's say I'm working on several different projects, where one set of Object Types only works for one of them, and makes no sense for the other project. In those cases it would be most intuitive that the maps themselves decides what is relevant object types and what is not. I could also see that this would be kinda frustrating for teams with more people involved, when they would have to load that objecttypes-file every time they switch a map.

From previous requests, as I understand, this is not something you are interested in implementing, which I think is too bad, since there obviously are a lot of people that would benefit from this feature. Would it be okay to have this feature if someone else implemented it? I would be interested to give it a try if that would be the only means to make this happen. 

I would suggest something like this, given the following example:
![image](https://user-images.githubusercontent.com/26206421/61125972-88067c00-a4ab-11e9-9088-d47263532229.png)
![image](https://user-images.githubusercontent.com/26206421/61126022-aa989500-a4ab-11e9-9f9b-3b4d35665acf.png)

You could have a property that is set on the map to make object types managed by the map itself.
In the exported tiled file (json as example):
![image](https://user-images.githubusercontent.com/26206421/61126166-0b27d200-a4ac-11e9-9806-fa6da345a22b.png)

With this property activated, we could store and tell Tiled that object types should be handled by the map itself. The data could be store like this:
![image](https://user-images.githubusercontent.com/26206421/61126823-cc931700-a4ad-11e9-9d79-a653c3b89f48.png)
Or as text format...

```javascript

...

"nextlayerid":3,
 "nextobjectid":85,
 "orientation":"orthogonal",
 "object_types":[
        {
            "name":"BlockType",
            "properties":[
                {
                 "name":"custom_property",
                 "type":"int",
                 "value":0
                }, 
                {
                 "name":"friction",
                 "type":"float",
                 "value":0.5
                },
                {
                    "name":"is_solid",
                    "type":"bool",
                    "value":true
                }]
        },
        {
            "name":"PlayerType",
            "properties":[
                {
                 "name":"base_hp",
                 "type":"int",
                 "value":3
                }, 
                {
                 "name":"name",
                 "type":"string",
                 "value":"James"
                }]
        }],
 "object_types_managed_by_map":true,
 "properties":[
        {
         "name":"camera_x",
         "type":"int",
         "value":-30
        }, 
        {
         "name":"camera_y",
         "type":"int",
         "value":0
        }],
 "renderorder":"left-down",
 "tiledversion":"1.2.3",
 "tileheight":16,
 "tilesets":[],
 "tilewidth":16,
 "type":"map",
 "version":1.2,
 "width":20

...

```

Could this be an acceptable feature? If it is something you really not want to implement, I could give it a try :) I think this would be very useful feature!

Sincerelly, 
Robin.

**Comment by bjorn** [comment]
> or at the very least have the possibility to export the actual properties of the Object Types from the Object Types Editor even when the value is the same as the default.

An option for this was added in Tiled 1.2. Under _Preferences > General > Export Options_ you can find _Resolve object types and properties_. What it will do is exactly like you describe, all inherited projects (either from tiles or from types) are then written explicitly to the exported map file.

Given that, do you still need an option to embed the actual object types in the map file?

**Comment by SSBMTonberry** [comment]
Hi, again!
I'm using Tiled 1.2.3, and I have the "Resolve object types and properties".
I might have explained myself a bit sloppy, but what I essentially would expect by what you describe was that I would get the properties of an object type exported directly to the map, but I cannot find and traces of them. Tried to export to both json and tmx, but it doesn't seem to display.

Let's say I'm using the object type PlayerType, from the initial post:
Then I would get two properties that I have inherited from that type: "base_hp" and "name":
![image](https://user-images.githubusercontent.com/26206421/61307296-5dd7f580-a7ee-11e9-893d-9c44901b89d7.png)

But I cannot find any traces to those properties in the exported file. I can see the type "PlayerType", but that doesn't do much unless I have that external file with the information. I would hope to have the properties "base_hp" and "name" (not to be confused with the tiled-property with the same name) there as well

This is what I find in my json-file:
```json
                {
                 "ellipse":true,
                 "height":16,
                 "id":133,
                 "name":"player",
                 "properties":[
                        {
                         "name":"color",
                         "type":"color",
                         "value":"#ffff0000"
                        }, 
                        {
                         "name":"shape",
                         "type":"string",
                         "value":"circle"
                        }],
                 "rotation":2,
                 "type":"PlayerType",
                 "visible":true,
                 "width":16,
                 "x":160,
                 "y":16
                }, 
```
Seeing this I am not completely sure what "Resolve object types and properties" is supposed to do.

If I could get those properties exported to my actual map, that would be good enough for me, but there would still be an issue when swapping environments (either computer, or some co-worker wanted to edit the map on their computer), unless you bring the external objecttypes file. Without the objecttypes file, the objecttypes in the map will not be valid anymore. This is where I think an option for having the object types controlled by the map itself would be the best solution (as I can think of anyway). That would make it much smoother to go from one map to another (let's say they are unrelated projects), and sharing maps between teams would be less of a headache.

In other words, I still would have much use for this feature. Is that something you would be willing to do, or at the very least would allow as a feature? I am positive to helping you out if it's too much of a hassle :)

**Comment by bjorn** [comment]
> But I cannot find any traces to those properties in the exported file.

When you say "exported file", do you really mean that you triggered the Export As action, or are you just talking about saving your file? The option in the Preferences only applies when you actually use the export action, since doing this on normal save is generally undesired (it would mark all those properties as "overridden").

> If I could get those properties exported to my actual map, that would be good enough for me, but there would still be an issue when swapping environments (either computer, or some co-worker wanted to edit the map on their computer), unless you bring the external objecttypes file. Without the objecttypes file, the objecttypes in the map will not be valid anymore.

Normally if you have a co-worker then you have a proper way to share files, for example using Dropbox or Syncthing or ideally a version control system. That's where you would put your `objecttypes.xml` file.

**Comment by SSBMTonberry** [comment]
Didn't know there was a difference between "Export As" and just saving the changes of the map, but you are correct: I get the expected properties when clicking "Export as".

With that said, I still think having an option to have property types managed by the map would be neat, but at least now I know there is an alternative where I can get those properties exported to the map, so I'm happy! Thanks :)

**Comment by bjorn** [comment]
Alright, glad we got that sorted at least! :-)

I'll close this issue since #228 is about putting this in the map. However, the general plan is to introduce support for projects (now scheduled for Tiled 1.4), which will also make it easier to keep the object types local to your project and share them with co-workers.
